hudi-client/src/main/java/org/apache/hudi/index/bloom/HoodieGlobalBloomIndex.java

```diff
@@ -98,26 +100,34 @@ public HoodieGlobalBloomIndex(HoodieWriteConfig config) {
       String recordKey = partitionRecordKeyPair._2();
       String partitionPath = partitionRecordKeyPair._1();
 
-      return indexFileFilter.getMatchingFiles(partitionPath, recordKey).stream()
-          .map(file -> new Tuple2<>(file, new HoodieKey(recordKey, indexToPartitionMap.get(file))))
+      return indexFileFilter.getMatchingFilesAndPartition(partitionPath, recordKey).stream()
+          .map(partitionFileIdPair -> new Tuple2<>(partitionFileIdPair.getRight(),
+              new HoodieKey(recordKey, partitionFileIdPair.getLeft())))
           .collect(Collectors.toList());
     }).flatMap(List::iterator);
   }
 
-
   /**
    * Tagging for global index should only consider the record key.
    */
   @Override
   protected JavaRDD<HoodieRecord<T>> tagLocationBacktoRecords(
       JavaPairRDD<HoodieKey, HoodieRecordLocation> keyFilenamePairRDD, JavaRDD<HoodieRecord<T>> recordRDD) {
-    JavaPairRDD<String, HoodieRecord<T>> rowKeyRecordPairRDD =
-        recordRDD.mapToPair(record -> new Tuple2<>(record.getRecordKey(), record));
 
-    // Here as the recordRDD might have more data than rowKeyRDD (some rowKeys' fileId is null),
-    // so we do left outer join.
-    return rowKeyRecordPairRDD.leftOuterJoin(keyFilenamePairRDD.mapToPair(p -> new Tuple2<>(p._1.getRecordKey(), p._2)))
-        .values().map(value -> getTaggedRecord(value._1, Option.ofNullable(value._2.orNull())));
+    Map<String, Pair<HoodieKey, HoodieRecordLocation>> keyMap = keyFilenamePairRDD
+        .mapToPair(e -> new Tuple2<>(e._1.getRecordKey(), Pair.of(e._1, e._2))).collectAsMap();
+
+    JavaRDD<HoodieRecord<T>> toReturn = recordRDD.map(rec -> {
+      if (keyMap.containsKey(rec.getRecordKey())) {
+        Pair<HoodieKey, HoodieRecordLocation> value = keyMap.get(rec.getRecordKey());
+        return getTaggedRecord(new HoodieRecord(value.getLeft(), rec.getData()), Option.of(value
```

Comment:
All we need to do is change the partitionPath in the tagged record right?

Reply:
That is what is happening here right? This logic only occurs when the record_key is already matching so we are essentially replacing just the partitionPath when we replace the HoodieKey.

Reply:
yes, you are right. keyMap contains existing record key information (HoodieKey, HoodieRecordLocation) in existing data set . For each incoming record to be upserted, if record key is located already, we need to upsert to the same partition even if incoming record (to be updated) was tagged with a different partition.And so, we reconstruct the HoodieKey with recordKey from incoming record and partition path from keyMap(index look up).

Reply:
@vinothchandar : yes, you are right.